File: python/3578.py
from collections import deque
from typing import List


class Solution:
    def countPartitions(self, nums: List[int], k: int) -> int:
        MOD = 10**9 + 7
        n = len(nums)
        dp = [0] * n
        prefix = [0] * n
        min_deque = deque()
        max_deque = deque()
        left = 0
        
        for i in range(n):
            num_i = nums[i]  # Cache to avoid repeated indexing
            
            # Maintain min deque (increasing order)
            while min_deque and nums[min_deque[-1]] >= num_i:
                min_deque.pop()
            min_deque.append(i)
            
            # Maintain max deque (decreasing order)
            while max_deque and nums[max_deque[-1]] <= num_i:
                max_deque.pop()
            max_deque.append(i)
            
            # Shrink window from left until valid
            while nums[max_deque[0]] - nums[min_deque[0]] > k:
                left += 1
                while min_deque and min_deque[0] < left:
                    min_deque.popleft()
                while max_deque and max_deque[0] < left:
                    max_deque.popleft()
            
            # Calculate dp[i]: sum of ways for all valid segments ending at i
            if left == 0:
                dp[i] = 1
            else:
                dp[i] = 0
            
            # Add sum of dp[j] for j from max(left-1, 0) to i-1
            if i > 0 and left <= i:
                start_idx = max(left - 1, 0)
                end_idx = i - 1
                if start_idx <= end_idx:
                    prefix_sum = prefix[end_idx]
                    if start_idx > 0:
                        prefix_sum = (prefix_sum - prefix[start_idx - 1] + MOD) % MOD
                    dp[i] = (dp[i] + prefix_sum) % MOD
            
            # Update prefix sum
            prefix[i] = ((prefix[i - 1] if i > 0 else 0) + dp[i]) % MOD
        
        return dp[n - 1]

File: cpp/3578.cpp
class Solution {
public:
    int countPartitions(vector<int>& nums, int k) {
        int n = nums.size();
        
        // dp[i] = number of ways to partition ending at index i
        vector<int> dp(n, 0);
        // prefix[i] = dp[0] + dp[1] + ... + dp[i]
        vector<int> prefix(n, 0);
        
        // Deques to maintain indices of min and max elements in sliding window
        deque<int> minDeque, maxDeque;
        
        int left = 0; // leftmost valid starting position for current window
        
        for (int i = 0; i < n; i++) {
            // maintain min deque (increasing order)
            while (!minDeque.empty() && nums[minDeque.back()] >= nums[i]) {
                minDeque.pop_back();
            }
            minDeque.push_back(i);
            
            // maintain max deque (decreasing order)
            while (!maxDeque.empty() && nums[maxDeque.back()] <= nums[i]) {
                maxDeque.pop_back();
            }
            maxDeque.push_back(i);
            
            // shrink window from left until it's valid
            while (nums[maxDeque.front()] - nums[minDeque.front()] > k) {
                left++;
                // remove indices that are out of window
                while (!minDeque.empty() && minDeque.front() < left) {
                    minDeque.pop_front();
                }
                while (!maxDeque.empty() && maxDeque.front() < left) {
                    maxDeque.pop_front();
                }
            }
            
            // now all segments from left to i are valid
            // dp[i] = sum of (j == 0 ? 1 : dp[j-1]) for j from left to i
            // This equals: (left == 0 ? 1 : 0) + sum(dp[j] for j from max(left-1, 0) to i-1)
            if (left == 0) {
                // segment starting at 0: add 1 way
                dp[i] = (dp[i] + 1) % MOD;
            }
            // Add sum of dp[j] for j from max(left-1, 0) to i-1
            if (i > 0 && left <= i) {
                int startIdx = max(left - 1, 0);
                int endIdx = i - 1;
                if (startIdx <= endIdx) {
                    int sum = (prefix[endIdx] - (startIdx > 0 ? prefix[startIdx - 1] : 0) + MOD) % MOD;
                    dp[i] = (dp[i] + sum) % MOD;
                }
            }
            
            // update prefix sum
            prefix[i] = ((i > 0 ? prefix[i - 1] : 0) + dp[i]) % MOD;
        }
        
        return dp[n - 1];
    }
private:
    static const int MOD = 1e9 + 7;
};